include/mqtt/endpoint.hpp

```diff
@@ -738,11 +741,14 @@ class endpoint : public std::enable_shared_from_this<endpoint<Mutex, LockGuard,
      */
     virtual void on_serialize_remove(packet_id_t packet_id) noexcept = 0;
 
+protected:
     /**
      * @brief Pre-send handler
      *        This handler is called when any mqtt control packet is decided to send.
      */
-    virtual void on_pre_send() noexcept = 0;
+    virtual void on_pre_send() noexcept {}
```

Comment:
Instead of making this an empty implementation, i would rather put an empty implementation into client.hpp as an override (but not final)

Reply:
same as https://github.com/redboltz/mqtt_cpp/pull/576#discussion_r364578346